mux.go

```diff
@@ -59,6 +59,12 @@ func (r *Router) Match(req *http.Request, match *RouteMatch) bool {
 			return true
 		}
 	}
+
+	// Closest match for a router (includes sub-routers)
+	if r.NotFoundHandler != nil {
+		match.Handler = r.NotFoundHandler
+		return true
```

Comment:
Or add a field to `RouteMatch` that says "this is a not-found handler, not an actual route match".
Or add a non-exported `matchHandler()` that does the regular `Match` behavior or returns `NotFoundHandler` if it's set.
I did not dig into the implementation of this package enough to come up with an actual valid suggestion, simply complaining :)

It basically feels like this is mixing up internal implementation details and exported API in a way that makes exported API unintuitive.